[SPARK-30776][ML] Support FValueRegressionSelector for continuous features and continuous labels

huaxingao opened the pull request:
### What changes were proposed in this pull request?
Add FValueRegressionSelector for continuous features and continuous labels.


### Why are the changes needed?
Current Spark only supports the selection of categorical features, while there are many requirements for the selection of continuous distribution features.

I will add two new selectors:

1. FValueRegressionSelector for continuous features and continuous labels.
2. ANOVAFValueClassificationSelector for continuous features and categorical labels.

I will use subtasks to add these two selectors:
1. add FValueRegressionSelector on scala side
2. add FValueRegressionSelector on python side
3. add samples and doc
4. do the same for ANOVAFValueClassificationSelector

### Does this PR introduce any user-facing change?
Yes.
Add a new Selector



### How was this patch tested?
Add new unit tests.

huaxingao committed:
[SPARK-30776][ML] Support FValueRegressionSelector for continuous features and continuous labels

SparkQA commented:
**[Test build #118173 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/118173/testReport)** for PR 27527 at commit [`d6b1626`](https://github.com/apache/spark/commit/d6b1626df150817484b0142a0c6c8bc8fcfe85ff).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22934/
Test PASSed.

SparkQA commented:
**[Test build #118173 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/118173/testReport)** for PR 27527 at commit [`d6b1626`](https://github.com/apache/spark/commit/d6b1626df150817484b0142a0c6c8bc8fcfe85ff).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FValueRegressionSelector(override val uid: String) extends`
  * `trait SelectionTestResult `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/118173/
Test FAILed.

huaxingao committed:
temporarily fix MimaExcludes to make test pass

SparkQA commented:
**[Test build #118174 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/118174/testReport)** for PR 27527 at commit [`2b519b5`](https://github.com/apache/spark/commit/2b519b5f1b5ffb02a285c80271892d3d0d826e76).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22935/
Test PASSed.

huaxingao commented on mllib/src/main/scala/org/apache/spark/ml/feature/Selector.scala:
```diff
@@ -0,0 +1,379 @@
+   * Only applicable when selectorType = "percentile".
+   * Default value is 0.1.
+   * @group param
+   */
```
In the original ChiSqSelector, the doc for percentile is this:
```
  /**
   * Percentile of features that selector will select, ordered by statistics value descending.
   * Only applicable when selectorType = "percentile".
   * Default value is 0.1.
   * @group param
   */
```
I changed it to ```ordered by ascending p-value```because in the below definition 
```
  final val percentile = new DoubleParam(this, "percentile",
    "Percentile of features that selector will select, ordered by ascending p-value.",
```
it has "ordered by ascending p-value". Also in the implementation of fit method, it uses pValue instead of statistics value.
```
      case "percentile" =>
        testResult
          .sortBy { case (res, _) => res.pValue }
          .take((testResult.length * getPercentile).toInt)
```

huaxingao commented on mllib/src/main/scala/org/apache/spark/ml/feature/Selector.scala:
```diff
@@ -0,0 +1,379 @@
+   * Only applicable when selectorType = "fpr".
+   * Default value is 0.05.
+   * @group param
+   */
```
Here is the original doc for ```fpr``` in ```ChiSqSelector```
```
  /**
   * The highest p-value for features to be kept.
   * Only applicable when selectorType = "fpr".
   * Default value is 0.05.
   * @group param
   */
```
I changed ```highest p-value``` to ```lowest p-value```

SparkQA commented:
**[Test build #118174 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/118174/testReport)** for PR 27527 at commit [`2b519b5`](https://github.com/apache/spark/commit/2b519b5f1b5ffb02a285c80271892d3d0d826e76).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/118174/
Test PASSed.

huaxingao commented:
cc @srowen @zhengruifeng

huaxingao commented:
### Outline of the changes:
1. Add new abstract classes Selector and SelectorModel. All the common code between ChiSqSelector and this newly added FValueRegressionSelector are put in these abstract classes. There are two abstract methods in Selector:
```
getSelectionTestResult(dataset: Dataset[_]): Array[SelectionTestResult]

createSelectorModel: T
```
2. Make ChiSqSelector extend Selector.  
implement ```getSelectionTestResult``` to return an array of ```ChiSqTestResult(pValue, degreeOfFreedom, statistics)  ```
pValue is used to rank the features and make selection
implement ```createSelectorModel``` to return a ```ChiSqSelectorModel```

3. FValueRegressionSelector extends Selector. 
implement ```getSelectionTestResult``` to return an array of ```
FValueRegressionTestResult(pValue, degreeOfFreedom, statistics)  
// statistics is fValue ```
pValue is used to rank the features and make selection
implement ```createSelectorModel``` to return a ```FValueRegressionSelectorModel```

```
fValue calculation:    X: feature      Y:label    N: numOfSample
degreeOfFreedom = N - 2
covariance = sum(((Xi - avg(X)) * ((Yi-avg(Y))) / (N-1)
correlation =  covariance / (Xstd * Ystd)
fValue = correlation * correlation / (1 - correlation * correlation) * degreeOfFreedom
```


4. The ChiSqSelectorModel constructor gets changed because two more parameters statistics and pValue were added. I think we should make all the SelectorModel (ChiSqSelectorModel and FRegressionSelectorModel) return statistics (chi square statistics or Fvalue)  and P-values. This is to address the comment in https://github.com/apache/spark/pull/27322.  
> I found that f_regression in scikit-learn will return both arrays of F-values and P-values, can we also add them to FRegressionSelectorModel? 


5. Because of adding two more parameters statistics and pValue in ChiSqSelectorModel constructor,  I added a ml-models/chisq-3.0.0 and modified the ChiSqSelectorSuite to make sure pre 3.1.0 model can be loaded OK in 3.1.0.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/stat/SelectionTest.scala:
```diff
@@ -0,0 +1,132 @@
+        (col, (value - xMeans(col.toInt), (label - yMean)))
+      }
+    }.aggregateByKey[(Double, Double)]((0.0, 0.0))(
+      seqOp = {
```
I suggest to use `E(XY)-E(X)E(Y)` instead, then only one pass is needed.

srowen replied:
Does this lead to numerical instability? i don't know either way, just know this has been an issue at times. Maybe it's 'worth it' here.

huaxingao replied:
I saw this on wikipedia (https://en.wikipedia.org/wiki/Covariance): 

![image](https://user-images.githubusercontent.com/13592258/74616568-6b838f80-50dd-11ea-9d5f-85c2416d244c.png)

Maybe we should use two pass algorithm instead of one pass?

zhengruifeng replied:
Yes, for numerical stablity

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/stat/SelectionTest.scala:
```diff
@@ -0,0 +1,132 @@
+
+    labeledPointRdd.flatMap { case LabeledPoint(label, features) =>
+      features.iterator.map { case (col, value) =>
+        (col, (value - xMeans(col.toInt), (label - yMean)))
```
I guess we only need to deal with non-zero values, so maybe use `foreachNonZero` instead.
Then I now think following the logic in `Summizer` maybe more efficient: maintain arrays of sum of `(value - xMeans(col.toInt)` and `(label - yMean)` on each partition, then `treeReduce` to obtain the global sum/mean.
I just notice that the FValueRegression's logic is different from ChiSqTest, since `ChiSqTest` need to maintain a relative large matrix for each col, so need to flatMap and aggByKey for each col. While in FValueRegression, we only need two arrays (or three arrays if using E(XY)-E(X)E(Y))

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/stat/SelectionTest.scala:
```diff
@@ -0,0 +1,132 @@
+    SchemaUtils.checkColumnType(dataset.schema, featuresCol, new VectorUDT)
+    SchemaUtils.checkNumericType(dataset.schema, labelCol)
+
+    val yMean = dataset.select(col(labelCol)).as[Double].rdd.stats().mean
```
Compute these variables on a pass
```
val Row(xMean: Vector, xStd: Vector, yMean: Double, yStd: Double, count: Long) = dataset
    .select(Summarizer.metrics("mean", "std", "count").summary(col($(featuresCol))).as("summary"),
            avg(col($(labelCol))).as("yMean"),
            stddev(col($(labelCol))).as("yStd"))
    .select("summary.mean", "summary.std", "yMean", "yStd", "summary.count")
    .first()
```

zhengruifeng replied:
Make sure the denominator is the same as `computeCovariance`,
if different, adjust it by multiplying `n/(n-1)` or `(n-1)/n` after converting back to variance.

zhengruifeng commented on mllib/src/main/scala/org/apache/spark/ml/stat/SelectionTest.scala:
```diff
@@ -0,0 +1,132 @@
+    var fTestResultArray = new Array[SelectionTestResult](numOfFeatures)
+
+    labeledPointRdd.flatMap { case LabeledPoint(label, features) =>
+      features.iterator.map { case (col, value) =>
```
Use array instead:
```scala
labeledPointRdd.mapPartition { iter =>
    if (iter.hasNext) {
        val array = Array.ofDim[Double](numFeatures)
        while(iter.hasNext) {
            val LabeledPoint(label, features) = iter.next
            val yDiff = label - yMean
            if (yDiff != 0) {
                features.iterator.zip(xMeans.iterator)
                    .foreach { case ((col, x), (_, xMean)) => array(col) += yDiff * (x - xMean) }
            }
        }
        Iterator.single(array)
    } else Iterator.empty
}.treeReduce { case (array1, array2) =>
    var i = 0
    while (i < numFeatures) {
        array1(i) += array2(i)
        i += 1
    }
    array1
}
```

zhengruifeng reviewed (commented):
In general, I am OK with it.
However, I think this PR is too large to track, I suggest to split it into three parts:
1, only add FValueRegression;
2, add ANOVAFValueSelector (not yet implemented in current PR);
3, add a common `Selector`, and make all three selector extend it.

huaxingao commented:
> However, I think this PR is too large to track, I suggest to split it into three parts...

I will close this PR and open smaller PRs. The comments are addressed in the new PR. Thanks a lot for reviewing!
Please see https://github.com/apache/spark/pull/27623
